File: whisper-drive-automation/src/runpod_client.py
"""
RunPod Serverless API client for Whisper transcription.

Handles:
- Job submission to RunPod endpoint
- Polling job status until completion
- Error handling and timeouts
"""

import requests
import time
import logging
from typing import Dict, Optional

logger = logging.getLogger(__name__)


class RunPodClient:
    """Client for RunPod Serverless API."""

    def __init__(self, api_key: str, endpoint: str):
        """
        Initialize RunPod client.

        Args:
            api_key: RunPod API key
            endpoint: RunPod endpoint URL (e.g., https://api.runpod.ai/v2/<ENDPOINT_ID>)
        """
        self.api_key = api_key
        self.endpoint = endpoint.rstrip('/')  # Remove trailing slash if present
        self.headers = {
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json"
        }

        logger.info(f"RunPod client initialized for endpoint: {endpoint}")

    def transcribe_audio(
        self,
        audio_url: str,
        model: str = "large-v3-turbo",
        language: str = "fr"
    ) -> Dict:
        """
        Send audio to RunPod for transcription.

        Args:
            audio_url: Publicly accessible URL to audio file
            model: Whisper model to use (default: large-v3-turbo)
            language: Language code (default: fr)

        Returns:
            Transcription result with segments and word-level timestamps

        Raises:
            requests.HTTPError: If API call fails
            TimeoutError: If job doesn't complete within timeout
            Exception: If job fails on RunPod side
        """
        logger.info(f"Submitting transcription job to RunPod (model={model}, language={language})")

        payload = {
            "input": {
                "audio": audio_url,
                "model": model,
                "language": language,
                "word_timestamps": True
            }
        }

        # Submit job to RunPod
        try:
            response = requests.post(
                f"{self.endpoint}/run",
                headers=self.headers,
                json=payload,
                timeout=30
            )
            response.raise_for_status()
        except requests.exceptions.RequestException as e:
            logger.error(f"Failed to submit RunPod job: {e}")
            raise

        result = response.json()
        job_id = result.get('id')

        if not job_id:
            raise Exception(f"RunPod did not return job ID. Response: {result}")

        logger.info(f"RunPod job submitted successfully. Job ID: {job_id}")

        # Poll for results
        return self._poll_job_status(job_id)

    def _poll_job_status(self, job_id: str, timeout: int = 1800, poll_interval: int = 5) -> Dict:
        """
        Poll job status until completion.

        Args:
            job_id: RunPod job ID
            timeout: Maximum time to wait in seconds (default: 1800 = 30 min)
            poll_interval: Time between status checks in seconds (default: 5)

        Returns:
            Job output/result

        Raises:
            TimeoutError: If job doesn't complete within timeout
            Exception: If job fails
        """
        start_time = time.time()
        elapsed = 0

        logger.info(f"Polling job {job_id} (timeout={timeout}s, interval={poll_interval}s)")

        while elapsed < timeout:
            try:
                response = requests.get(
                    f"{self.endpoint}/status/{job_id}",
                    headers=self.headers,
                    timeout=30
                )
                response.raise_for_status()
            except requests.exceptions.RequestException as e:
                logger.warning(f"Error checking job status: {e}. Retrying...")
                time.sleep(poll_interval)
                elapsed = time.time() - start_time
                continue

            data = response.json()
            status = data.get('status')

            if status == 'COMPLETED':
                logger.info(f"Job {job_id} completed successfully")
                output = data.get('output')

                if not output:
                    raise Exception(f"Job completed but no output returned. Response: {data}")

                return output

            elif status == 'FAILED':
                error_msg = data.get('error', 'Unknown error')
                logger.error(f"Job {job_id} failed: {error_msg}")
                raise Exception(f"RunPod job failed: {error_msg}")

            elif status in ['IN_QUEUE', 'IN_PROGRESS']:
                elapsed = time.time() - start_time
                logger.debug(f"Job {job_id} status: {status} ({elapsed:.1f}s elapsed)")

            else:
                logger.warning(f"Unknown job status: {status}")

            time.sleep(poll_interval)
            elapsed = time.time() - start_time

        # Timeout reached
        logger.error(f"Job {job_id} timed out after {timeout}s")
        raise TimeoutError(f"RunPod job {job_id} timed out after {timeout}s")

    def cancel_job(self, job_id: str) -> bool:
        """
        Cancel a running job.

        Args:
            job_id: RunPod job ID

        Returns:
            True if cancellation successful
        """
        try:
            response = requests.post(
                f"{self.endpoint}/cancel/{job_id}",
                headers=self.headers,
                timeout=30
            )
            response.raise_for_status()
            logger.info(f"Job {job_id} cancelled successfully")
            return True
        except requests.exceptions.RequestException as e:
            logger.error(f"Failed to cancel job {job_id}: {e}")
            return False

File: transcription-desktop-app/src/services/runpod-client.js
'use strict';

/**
 * RunPod Serverless API client for Whisper transcription.
 * Port of runpod_client.py
 */

const https = require('https');
const http = require('http');

class RunPodClient {
  constructor(apiKey, endpoint) {
    this.apiKey = apiKey;
    this.endpoint = endpoint.replace(/\/$/, '');
    this.headers = {
      'Authorization': `Bearer ${apiKey}`,
      'Content-Type': 'application/json'
    };
  }

  /**
   * Submit audio URL to RunPod and poll until result.
   * @param {string} audioUrl - Public URL to audio file
   * @param {string} model - Whisper model (default: turbo)
   * @param {string} language - Language code (default: fr)
   * @returns {Promise<object>} Transcription result
   */
  async transcribeAudio(audioUrl, model = 'turbo', language = 'fr') {
    const payload = {
      input: {
        audio: audioUrl,
        model,
        language,
        word_timestamps: true,
        condition_on_previous_text: false,
        no_speech_threshold: 0.8,
        compression_ratio_threshold: 3.0
      }
    };

    const result = await this._request('POST', `${this.endpoint}/run`, payload);
    const jobId = result.id;

    if (!jobId) {
      throw new Error(`RunPod did not return job ID. Response: ${JSON.stringify(result)}`);
    }

    return this._pollJobStatus(jobId);
  }

  /**
   * Poll job status until COMPLETED or FAILED.
   * @param {string} jobId
   * @param {number} timeout - Max wait in seconds (default 1800)
   * @param {number} pollInterval - Poll interval in seconds (default 5)
   */
  async _pollJobStatus(jobId, timeout = 1800, pollInterval = 5) {
    const start = Date.now();
    let consecutiveErrors = 0;
    const maxRetries = 3;

    while ((Date.now() - start) / 1000 < timeout) {
      let data;
      try {
        data = await this._request('GET', `${this.endpoint}/status/${jobId}`);
        consecutiveErrors = 0;
      } catch (err) {
        consecutiveErrors++;
        if (consecutiveErrors >= maxRetries) {
          throw new Error(`Failed to check job status after ${maxRetries} retries: ${err.message}`);
        }
        await this._sleep(pollInterval * 1000);
        continue;
      }

      const status = data.status;

      if (status === 'COMPLETED') {
        if (!data.output) {
          throw new Error(`Job completed but no output returned. Response: ${JSON.stringify(data)}`);
        }
        return data.output;
      } else if (status === 'FAILED') {
        throw new Error(`RunPod job failed: ${data.error || 'Unknown error'}`);
      }
      // IN_QUEUE or IN_PROGRESS → keep polling

      await this._sleep(pollInterval * 1000);
    }

    throw new Error(`RunPod job ${jobId} timed out after ${timeout}s`);
  }

  /**
   * Cancel a running job.
   */
  async cancelJob(jobId) {
    try {
      await this._request('POST', `${this.endpoint}/cancel/${jobId}`);
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Generic HTTP request helper (no external dependencies).
   */
  _request(method, url, body = null) {
    return new Promise((resolve, reject) => {
      const parsedUrl = new URL(url);
      const transport = parsedUrl.protocol === 'https:' ? https : http;

      const bodyStr = body ? JSON.stringify(body) : null;
      const options = {
        hostname: parsedUrl.hostname,
        port: parsedUrl.port || (parsedUrl.protocol === 'https:' ? 443 : 80),
        path: parsedUrl.pathname + parsedUrl.search,
        method,
        headers: {
          ...this.headers,
          ...(bodyStr ? { 'Content-Length': Buffer.byteLength(bodyStr) } : {})
        },
        timeout: 30000
      };

      const req = transport.request(options, (res) => {
        let raw = '';
        res.on('data', (chunk) => { raw += chunk; });
        res.on('end', () => {
          if (res.statusCode >= 400) {
            return reject(new Error(`HTTP ${res.statusCode}: ${raw}`));
          }
          try {
            resolve(JSON.parse(raw));
          } catch {
            resolve(raw);
          }
        });
      });

      req.on('error', reject);
      req.on('timeout', () => { req.destroy(); reject(new Error('Request timeout')); });

      if (bodyStr) req.write(bodyStr);
      req.end();
    });
  }

  _sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}

module.exports = RunPodClient;
